add memory access patch for QEMU v2.8.0

Wenzel opened the pull request:
This patch updates the memory access QMP command for QEMU v2.8.0.
- `qmp-commands.hx` has been replaced by `docs/qmp-commands.txt` since this [commit](https://github.com/qemu/qemu/commit/bd6092e407a5d682fbfddcbc510038b84bc1a1aa)
- tested with latest libvmi using the `process-list` example

libvmi-jenkins commented:
Can one of the admins verify this patch?

tklengyel commented on tools/qemu-kvm-patch/qemu-2.8-memaccess.patch:
```diff
@@ -0,0 +1,362 @@
++# the socket as:
++#
++# struct request {
++#     uint8_t type;   // 0 quit, 1 read, 2 write, ... rest reserved
```
So since this is an ABI I think it would be much better if we changed this to a uint64_t as well as to avoid various compilers deciding on different padding after it. I realize we have been using this in all other patches too but at least going forward we should be safer.

Wenzel replied:
good idea. i just pushed the fix, but what about using `__attribute__((packed)) ` ?
Does all compilers use the same alignement strategy when there is this attribute specified ?

tklengyel replied:
`__attribute__((packed))` is a GCC specific C extension, so other compilers may not honor it

Wenzel replied:
At least, it works with Clang also.
But let's go for replacing with `uint64_t`

tklengyel commented:
```
/root/qemu-2.8.0/memory-access.c: In function ‘connection_handler’:
/root/qemu-2.8.0/memory-access.c:135:13: warning: format ‘%d’ expects argument of type ‘int’, but argument 3 has type ‘uint64_t’ [-Wformat=]
             fprintf(stderr, "Qemu pmemaccess: ignoring unknown command (%d)\n", req.type);
             ^
  CC      aarch64-softmmu/qtest.o
  CC      alpha-softmmu/memory-access.o
/root/qemu-2.8.0/memory-access.c: In function ‘connection_handler’:
/root/qemu-2.8.0/memory-access.c:135:13: warning: format ‘%d’ expects argument of type ‘int’, but argument 3 has type ‘uint64_t’ [-Wformat=]
             fprintf(stderr, "Qemu pmemaccess: ignoring unknown command (%d)\n", req.type);
             ^
  CC      alpha-softmmu/qtest.o

tklengyel commented:
This is ok to test.

tklengyel commented:
The `struct request` definition in LibVMI driver/kvm/kvm.c is now out-of-sync with what is being defined in this patch. IMHO it would be best if the structure wouldn't be defined in an exported header that LibVMI can include and only define it locally if no other definition was found.

Wenzel commented:
So `LibVMI` should install that header ?
Or the modified `QEMU` should do it ?

tklengyel commented:
The modified QEMU should export that header.

tklengyel commented on configure.ac:
```diff
@@ -66,6 +66,11 @@ YES_IS_DEFINED
 dnl -----------------------------------------------
 dnl Check package options
 dnl -----------------------------------------------
+ AC_CHECK_HEADER([qemu/libvmi_request.h],
```
This check should be moved below to the `[if test "$enable_kvm" = "yes"]` block.

tklengyel replied:
Ping.

Wenzel commented:
I'm more a fan of `CMake` vs `autotools`, so i copy pasted this `AC_CHECK_HEADER` from stackoverflow :)

If there is a more elegant way of writing it, and better, if you want to display a hint about `libvmi_request.h` in the `configure` summary, go ahead !

tklengyel commented:
The check itself is fine, it's position is what is still not good. There is a block of checks that only get performed if KVM is enabled. This check should be in the same block.

tklengyel commented on configure.ac:
```diff
@@ -217,6 +217,10 @@ AC_SUBST([GLIB_LIBS])
     AC_CHECK_HEADER(libvirt/libvirt.h, [],
         [AC_MSG_ERROR([No libvirt headers found. Install missing package or re-run with --disable-kvm])])
     AC_DEFINE([ENABLE_KVM], [1], [Define to 1 to enable KVM support.])
+    AC_CHECK_HEADER([qemu/libvmi_request.h],
```
Almost there, just has to move it before the AC_DEFINE line =)

tklengyel commented:
Retest this please.

Wenzel commented:
i just rebased it into one commit.

Wenzel commented:
Can you explain what caused the test fail for `PR_LibVMI_KVM_Test_Ubuntu1204-64` ?

tklengyel commented:
Retest this please.

tklengyel commented:
So I've updated the KVM test machine to use QEMU 2.8 with this patch. It spins at 100% and then times out.

```                                                          
   803 libvirt+  20   0 2274452 580436  21720 R  99.5 28.8   5:03.03 qemu-system-x86

tklengyel commented:
Retest this please.

tklengyel commented:
Retest this please.

tklengyel commented:
OK, so the problem had been that QEMU 2.8 is installed in `/opt`, thus ./configure wasn't finding `qemu/libvmi_request.h` properly. LibVMI then was silently compiled with the legacy struct definition, which of course hung the VM. To avoid this issue it might be worthwhile to explore adding an extra check (compile- or run-time) to verify that if we are running on QEMU 2.8+ that this struct was defined from the header.

Wenzel commented:
How do you propose to implement that check ?
Maybe adding the hint in the configure summary would be the most simple thing to do here.

tklengyel commented:
Doing it in configure is not necessarily the right place because you could compile LibVMI on one machine but run it on another. Hence a runtime check would be more robust. Though I'm not sure how that would be done.

Wenzel commented:
What about adding a version number at the end of the `request` struct ?

Therefore we can check at runtime when we try to initalize the `KVM` driver.
On a standard QEMU `2.8`, the packet sent on the socket will be too small, and the `read(kvm_get_instance(vmi)->socket_fd, buf, length + 1);` will return an error.

On a QEMU `2.8+`, we can compare the version number against a defined constant in `libvmi_request.h`

tklengyel commented:
Adding versioning to the struct is a good idea but relying on the out-of-bound read to fail is not a nice way to signal a version-mismatch. There really is no existing API to get the qemu version at runtime?

Wenzel commented:
Ah, sorry, i didn't really understood when you mentionned QEMU `2.8+`.
So yes, there is the montior command `info` with the subcommand `version`
~~~
(qemu) info version
2.6.2 (qemu-2.6.2-7.fc24)
~~~

I can append some extra info there, and we could parse it when initalizing the driver.

Wenzel commented:
So, the QEMU version is defined in the `VERSION` file, at the root of the repo :
`VERSION`
~~~
2.8.0
~~~

I tried to append `-vmi` here, however, it is not parsed as a string internally, but as 3 integers(https://github.com/qemu/qemu/blob/master/qmp.c#L53).

So as i didn't wanted to change a lot of things about version parsing inside QEMU, i just modified the `hmp_version` [function to append `vmi` for now](https://github.com/KVM-VMI/qemu/blob/qemu-v2.8-pmemaccess/hmp.c#L73)
~~~C
void hmp_info_version(Monitor *mon, const QDict *qdict)
{
    VersionInfo *info;

    info = qmp_query_version(NULL);

    monitor_printf(mon, "%" PRId64 ".%" PRId64 ".%" PRId64 "%s %s\n",
                   info->qemu->major, info->qemu->minor, info->qemu->micro,
                   info->package,
                   "vmi");

    qapi_free_VersionInfo(info);
}
~~~

On the libvmi KVM driver side, i added this runtime check in `kvm_init_vmi`
~~~C
#ifdef HAVE_LIBVMI_REQUEST
    char *version = exec_info_version(kvm);
    dbprint(VMI_DEBUG_KVM, "Checking QEMU version string : %s", version);
    // find substring "vmi"
    if (NULL == strstr(version, "vmi"))
    {
        dbprint(VMI_DEBUG_KVM, "Failed find 'vmi' in QEMU version\n");
        return VMI_FAILURE;
    }
#endif
~~~

I tested with `./vmi-process-list win7x64` in `examples`, it works.

Here is the debug output:
~~~
--qmp: virsh -c qemu:///system qemu-monitor-command nitro_win7x64 '{"execute": "human-monitor-command", "arguments": {"command-line": "info version"}}'
Checking QEMU version string : {"return":"2.8.0 vmi\r\n","id":"libvirt-40"}
~~~

Tell me what you think.
- should i move the runtime check somewhere else ?
- should i add more debug statements ?

tklengyel commented on libvmi/driver/kvm/kvm.c:
```diff
@@ -1364,6 +1380,17 @@ kvm_init_vmi(
     }
     vmi->num_vcpus = info.nrVirtCpu;
 
+#ifdef HAVE_LIBVMI_REQUEST
```
The whole point of checking the QEMU version is to determine if not having the request structure defined by a header is a problem or not. So right now you have it backwards. If the pre-processor check HAVE_LIBVMI_REQUEST succeeds we know we are 2.8+ and have the libvmi header. What we need is to check the runtime version and if it is 2.8+ and we don't have HAVE_LIBVMI_REQUEST then print an error message and exit.

Wenzel commented:
I reversed the runtime check, but for me this looks like the same at the end =)

tklengyel commented on libvmi/driver/kvm/kvm.c:
```diff
@@ -1364,6 +1380,17 @@ kvm_init_vmi(
+#ifndef HAVE_LIBVMI_REQUEST
+    char *version = exec_info_version(kvm);
+    dbprint(VMI_DEBUG_KVM, "--Checking QEMU version string : %s", version);
+    // find substring "vmi"
```
So there is no point appending "vmi" to the version string. The version should be checked whether it is 2.8 or later and if so, error out.

Wenzel replied:
Ok, so i will remove `vmi` to the QEMU version string.
And to check that the QEMU version is `2.8.0` or above, we need to parse it.

As i would like to avoid manually parsing a string in C, because it's error prone and painful, i would rather use the QMP command `query-version` and load the json output:
~~~JSON
{"return":{"qemu":{"micro":0,"minor":8,"major":2},"package":""},"id":"libvirt-35"}
~~~
which will allow me to have the `major`, `minor` and `micro` int version number directly.
This would add a dependency on a JSON library.

Do we already use a JSON parser in libvmi ?
What do you think of this solution ?

tklengyel replied:
That works for me. There is a json-c dependency for Rekall profiles already, so adding that as a KVM dependency too would be fine.

Wenzel replied:
I may have noticed an issue in where we are doing the check.

the KVM driver has 2 memory accesses:
- the slow one, using QEMU commands
- the fast one, using the socket and the modified QEMU

Here, we are always failing if we detect that QEMU is `v2.8.0` or later.
So we deny the possibility of using the slow memory access.

Is it what you want, and focus only on the modified QEMU ?
Did i misunderstood something ?

Wenzel commented:
Any update ? :)

tklengyel commented:
Retest this please.

Wenzel commented:
Hi @tklengyel ,

Just to know the status of this PR, are you waiting for me to make some changes, or is it OK for your review ?

Thanks :)

Wenzel commented:
The test `PR_LibVMI_KVM_Test_Winxpsp2` fails because it cannot find the C JSON library

~~~
libvmi/include -MT driver/kvm/libvmi_la-kvm.lo -MD -MP -MF driver/kvm/.deps/libvmi_la-kvm.Tpo -c driver/kvm/kvm.c  -fPIC -DPIC -o driver/kvm/.libs/libvmi_la-kvm.o
driver/kvm/kvm.c:45:25: fatal error: json-c/json.h: No such file or directory
 #include <json-c/json.h>
                         ^
compilation terminated.
~~~

tklengyel commented:
Retest this please.

tklengyel commented:
@Wenzel Looks good now! Thanks =)

tklengyel commented:
Retest this please.
